hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -696,3 +696,73 @@ def valid_tuple_axes(ndim, min_size=0, max_size=None):
         lambda x: x if x < ndim else x - 2 * ndim
     )
     return st.lists(axes, min_size, max_size, unique_by=lambda x: x % ndim).map(tuple)
+
+
+def _insert_singleton(mask, shape):
+    # type: (Sequence[bool], Sequence[int]) -> List[int]
+    return [1 if m else s for m, s in zip(mask, reversed(shape))][::-1]
+
+
+def _aligned_shapes(shape, min_dims, max_dims):
+    # type: (Sequence[int], int, int) -> st.SearchStrategy[List[int]]
+    return st.lists(st.booleans(), min_size=min_dims, max_size=max_dims).map(
+        lambda x: _insert_singleton(x, shape)
+    )
+
+
+@st.composite
+def _broadcastable_shapes(draw, shape, min_dims=0, max_dims=3, min_side=0, max_side=5):
+    if max_dims <= len(shape):
+        return tuple(draw(_aligned_shapes(shape, min_dims, max_dims)))
+
+    max_lead = max_dims - len(shape)
+    min_lead = max(min_dims - len(shape), 0)
+    leading = draw(
+        st.lists(st.integers(min_side, max_side), min_size=min_lead, max_size=max_lead)
+    )
+    min_aligned_dim = len(shape) if leading else min_dims
+    aligned = draw(_aligned_shapes(shape, min_dims=min_aligned_dim, max_dims=max_dims))
+    return tuple(leading + aligned)
+
+
+@st.defines_strategy
+def broadcastable_shapes(shape, min_dims=0, max_dims=3, min_side=1, max_side=5):
+    # type: (Sequence[int], int, int, int, int) -> st.SearchStrategy[Tuple[int, ...]]
+    """Return a strategy for generating shapes that are broadcast-compatible
+    with the provided shape.
+
+    Examples from this strategy shrink towards a shape with length ``min_dims``.
+    The size of an aligned dimension shrinks away from being a singleton. The
+    size of an unaligned dimension shrink towards ``min_side``.
+
+    * ``shape`` a sequence of integers
```

Comment:
Let's just be super-strict and require this to be a `tuple` of integers; as-is the docs and type hints allow any sequence but at runtime only lists or tuples are allowed - and the contents are not checked yet.